rasa/engine/caching.py

```diff
@@ -0,0 +1,373 @@
+        ]
+        with self._sessionmaker.begin() as session:
+            delete_query = sa.delete(self.CacheEntry).where(
+                self.CacheEntry.fingerprint_key.in_(incompatible_fingerprints)
+            )
+            session.execute(delete_query)
+
+    @staticmethod
+    def _delete_cached_result(entry: TrainingCache.CacheEntry) -> None:
+        if entry.result_location and Path(entry.result_location).is_dir():
+            shutil.rmtree(entry.result_location)
+
+    def cache_output(
+        self,
+        fingerprint_key: Text,
+        output: Any,
+        output_fingerprint: Text,
+        model_storage: ModelStorage,
+    ) -> None:
+        """Adds the output to the cache.
+
+        If the output is of type `Cacheable` the output is persisted to disk in addition
+        to its fingerprint.
+
+        Args:
+            fingerprint_key: The fingerprint key serves as key for the cache. Graph
+                components can use their fingerprint key to lookup fingerprints of
+                previous training runs.
+            output: The output. The output is only cached to disk if it's of type
+                `Cacheable`.
+            output_fingerprint: The fingerprint of they output. This can be used
+                to lookup potentially persisted outputs on disk.
+            model_storage: Required for caching `Cacheable` instances. E.g. `Resource`s
+                use that to copy data from the model storage to the cache.
+        """
+        if self._is_disabled():
+            return
+
+        cache_dir, output_type = None, None
+        if isinstance(output, Cacheable):
+            cache_dir, output_type = self._cache_output_to_disk(output, model_storage)
+
+        with self._sessionmaker.begin() as session:
+            cache_entry = self.CacheEntry(
+                fingerprint_key=fingerprint_key,
+                output_fingerprint_key=output_fingerprint,
+                last_used=datetime.utcnow(),
+                rasa_version=rasa.__version__,
+                result_location=cache_dir,
+                result_type=output_type,
+            )
+            session.add(cache_entry)
+
+    def _is_disabled(self) -> bool:
+        return self._max_cache_size == 0.0
+
+    def _cache_output_to_disk(
+        self, output: Cacheable, model_storage: ModelStorage
+    ) -> Tuple[Optional[Text], Optional[Text]]:
+        with tempfile.TemporaryDirectory() as temp_dir:
+            try:
+                output.to_cache(Path(temp_dir), model_storage)
+
+                logger.debug(f"Caching output of type '{type(output)}' succeeded.")
+            except Exception as e:
+                logger.error(
+                    f"Caching output of type '{type(output)}' failed with the "
+                    f"following error:\n{e}"
+                )
+                return None, None
+
+            output_size = _directory_size_in_mb(temp_dir)
+            if output_size > self._max_cache_size:
```

Comment:
I wonder if we also want a `max_cache_entry_size`?

Reply:
what added benefit would it provide?

Reply:
I guess the question is, do we want to cache something which fills the entire cache?
Another alternative would be to allow a certain proportion of the max cache size. But maybe not needed.

Reply:
I'd go for the simple LRU mechanism without any secret sauce to it for now. I am explicitly excluding the database from the size calculation now and added some handling for potential infinite loops